src/WinRT.Runtime/ComWrappersSupport.net5.cs

```diff
@@ -99,12 +105,22 @@ public static T CreateRcwForComObject<T>(IntPtr ptr)
                 winrtObj.Resurrect();
             }
 
-            return rcw switch
-            {
-                ABI.System.Nullable<string> ns => (T)(object) ns.Value,
-                ABI.System.Nullable<Type> nt => (T)(object) nt.Value,
-                _ => (T) rcw
-            };
+            try
+            {
+                return rcw switch
+                {
+                    ABI.System.Nullable<string> ns => (T)(object)ns.Value,
+                    ABI.System.Nullable<Type> nt => (T)(object)nt.Value,
+                    _ => (T)rcw
+                };
+            }
+            catch (InvalidCastException)
+            {
+                if (tryUseCache)
+                    return CreateRcwForComObject<T>(ptr, false);
+                throw;
+            }

```

Comment:
I'd suggest using a pattern matching mechanism to avoid having any `InvalidCastException`s in a non-failure path.

```suggestion
            return rcw switch
            {
                ABI.System.Nullable<string> ns => (T)(object)ns.Value,
                ABI.System.Nullable<Type> nt => (T)(object)nt.Value,
                T castRcw => castRcw
                _ when tryUseCache => CreateRcwForComObject<T>(ptr, false),
                _ => throw new InvalidCastException()
            };
```